apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl

```diff
@@ -0,0 +1,505 @@
+    {GenId, Gen} = meta_lookup_gen(Shard, StartTime),
+    open_iterator(Gen, #it{
+        shard = Shard,
+        gen = GenId,
+        replay = Replay
+    }).
+
+-spec next(iterator()) -> {value, binary(), iterator()} | none | {error, closed}.
+next(It = #it{module = Mod, data = ItData}) ->
+    case Mod:next(ItData) of
+        {value, Val, ItDataNext} ->
+            {value, Val, It#it{data = ItDataNext}};
+        {error, _} = Error ->
+            Error;
+        none ->
+            case open_next_iterator(It) of
+                {ok, ItNext} ->
+                    next(ItNext);
+                {error, _} = Error ->
+                    Error;
+                none ->
+                    none
+            end
+    end.
+
+-spec preserve_iterator(iterator(), emqx_ds:replay_id()) ->
+    ok | {error, _TODO}.
+preserve_iterator(It = #it{}, ReplayID) ->
+    iterator_put_state(ReplayID, It).
+
+-spec restore_iterator(emqx_ds:shard(), emqx_ds:replay_id()) ->
+    {ok, iterator()} | {error, _TODO}.
+restore_iterator(Shard, ReplayID) ->
+    case iterator_get_state(Shard, ReplayID) of
+        {ok, Serial} ->
+            restore_iterator_state(Shard, Serial);
+        not_found ->
+            {error, not_found};
+        {error, _Reason} = Error ->
+            Error
+    end.
+
+-spec discard_iterator(emqx_ds:shard(), emqx_ds:replay_id()) ->
+    ok | {error, _TODO}.
+discard_iterator(Shard, ReplayID) ->
+    iterator_delete(Shard, ReplayID).
+
+%%================================================================================
+%% behaviour callbacks
+%%================================================================================
+
+init([Shard]) ->
+    process_flag(trap_exit, true),
+    {ok, S0} = open_db(Shard),
+    S = ensure_current_generation(S0),
+    ok = populate_metadata(S),
+    {ok, S}.
+
+handle_call({create_generation, Since, Config}, _From, S) ->
+    case create_new_gen(Since, Config, S) of
+        {ok, GenId, NS} ->
+            {reply, {ok, GenId}, NS};
+        {error, _} = Error ->
+            {reply, Error, S}
+    end;
+handle_call(_Call, _From, S) ->
+    {reply, {error, unknown_call}, S}.
+
+handle_cast(_Cast, S) ->
+    {noreply, S}.
+
+handle_info(_Info, S) ->
+    {noreply, S}.
+
+terminate(_Reason, #s{db = DB, shard = Shard}) ->
+    meta_erase(Shard),
+    ok = rocksdb:close(DB).
+
+%%================================================================================
+%% Internal functions
+%%================================================================================
+
+-record(db, {handle :: rocksdb:db_handle(), cf_iterator :: rocksdb:cf_handle()}).
+
+-spec populate_metadata(state()) -> ok.
+populate_metadata(S = #s{shard = Shard, db = DBHandle, cf_iterator = CFIterator}) ->
+    ok = meta_put(Shard, db, #db{handle = DBHandle, cf_iterator = CFIterator}),
+    Current = schema_get_current(DBHandle),
+    lists:foreach(fun(GenId) -> populate_metadata(GenId, S) end, lists:seq(0, Current)).
```

Comment:
Q: this seems to potentially set the `current` gen multiple times for a shard while it's starting; since `persistent_term` may perform global GC for updates, wouldn't it be better to set the current generation after all generation metadata keys have been populated?

Reply:
Overall, yes it would. However, `current` gen here is just a (short) integer, persistent term GC for such updates should be no-op.